File: dotnet/Functions/SendOtp.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.Logging;

namespace Epp.Otp;

// Echo the nonce only on acceptance; keep service events and the request summary PII-safe.
public sealed class SendOtp
{
    private readonly DispatchEngine _engine;
    private readonly JweDecryptor _decryptor;
    private readonly IEnv _env;
    private readonly ILogger<SendOtp> _log;

    public SendOtp(DispatchEngine engine, JweDecryptor decryptor, IEnv env, ILogger<SendOtp> log)
    {
        _engine = engine;
        _decryptor = decryptor;
        _env = env;
        _log = log;
    }

    // Anonymous at the Functions layer; EasyAuth must remain enabled and require authentication in the cloud.
    [Function("SendOtp")]
    public async Task<IActionResult> Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "SendOtp")] HttpRequest req,
        FunctionContext? functionContext = null)
    {
        var requestId = Guid.NewGuid().ToString("n");
        var msRequestId = req.Headers["x-ms-client-request-id"].FirstOrDefault();
        var headerCorrelationId = req.Headers["x-ms-correlation-id"].FirstOrDefault();
        var log = new RequestLog(_log, requestId, functionContext?.InvocationId, msRequestId, headerCorrelationId);
        var correlationId = headerCorrelationId ?? requestId;
        var httpStatus = 500;
        var evaluation = false;

        ObjectResult Reply(int status, object body)
        {
            var response = new ObjectResult(body) { StatusCode = status };
            httpStatus = status;
            log.ResponsePrepared(status, body is EndpointSuccessResponse,
                body is EndpointSuccessResponse or EndpointErrorResponse { CorrelationId: not null });
            return response;
        }

        try
        {
            log.Service("request_received");
            var config = AppConfig.Read(_env);
            var clientRequestId = msRequestId ?? requestId;

            var (envelope, envelopeError) = await EnvelopeParser.ParseAsync(req.Body, req.HttpContext.RequestAborted);
            if (envelopeError is not null)
            {
                log.Failure("request_validation", envelopeError, 400);
                return Reply(400, new EndpointErrorResponse("bad_request", requestId, Reason: envelopeError));
            }

            correlationId = envelope!.CorrelationId ?? correlationId;
            evaluation = envelope.Mode == EnvelopeParser.ModeEvaluation;
            log.EnvelopeValidated(envelope, envelope.CorrelationId ?? headerCorrelationId,
                envelope.CorrelationId is not null ? "envelope" : "header");

            JweResult decrypted;
            try
            {
                decrypted = _decryptor.Decrypt(envelope.EncryptedDeliveryContext);
            }
            catch
            {
                log.Failure("decryption", "decryption_failed", 400);
                return Reply(400, new EndpointErrorResponse("decryption_failed", requestId, CorrelationId: correlationId));
            }
            log.Service("delivery_context_decrypted");

            if (!string.IsNullOrEmpty(config.ExpectedKeyId)
                && !string.Equals(config.ExpectedKeyId, decrypted.Kid, StringComparison.Ordinal))
                log.KeyIdMismatch();

            var context = decrypted.Context;
            if (!context.IsComplete)
            {
                log.Failure("delivery_context_validation", "incomplete delivery context", 400);
                return Reply(400, new EndpointErrorResponse("bad_request", requestId, Reason: "incomplete delivery context", CorrelationId: correlationId));
            }

            // Evaluation proves validation/decryption without requiring any provider configuration.
            if (evaluation)
            {
                log.Service("evaluation_completed");
                return Reply(200, new EndpointSuccessResponse(context.Nonce!, correlationId));
            }

            var channel = EnvelopeParser.ChannelName(envelope.Channel)!;

            var dispatch = new DispatchRequest(
                Destination: context.PhoneNumber!,
                Message: context.Message,
                Channel: channel,
                MessageId: clientRequestId,
                CorrelationId: correlationId,
                Locale: context.Locale,
                TextToVoice: context.TextToVoice);

            // A nonce acknowledges delivery, not just decryption. Wait for the bounded provider call.
            var result = await _engine.DispatchAsync(dispatch, requestId, log);
            if (result.HttpStatus != 200)
                return Reply(result.HttpStatus, new EndpointErrorResponse("provider_delivery_failed", requestId, CorrelationId: correlationId));

            return Reply(200, new EndpointSuccessResponse(context.Nonce!, correlationId));
        }
        catch
        {
            if (!log.HasFailure) log.Failure("handler", "unexpected_error", 500);
            return Reply(500, new EndpointErrorResponse("delivery_failed", requestId, CorrelationId: correlationId));
        }
        finally
        {
            log.Complete(httpStatus);
        }
    }
}

File: javascript/src/functions/SendOtp.js
// <copyright file="SendOtp.js" company="Microsoft Corporation">
// Copyright (c) Microsoft Corporation. All rights reserved.
// </copyright>

'use strict';

const { app } = require('@azure/functions');
const crypto = require('crypto');
const {
    dispatchOtp,
    parseEnvelope,
    decryptDeliveryContext,
    contextToDispatch,
    startProviderCredentialRefresh,
    stopProviderCredentialRefresh,
    MODE,
} = require('./dispatch');
const { readConfig } = require('./config');
const { RequestLog } = require('./requestLog');

app.hook.appStart(startProviderCredentialRefresh);
app.hook.appTerminate(stopProviderCredentialRefresh);

app.http('SendOtp', {
    methods: ['POST'],
    authLevel: 'anonymous', // Protected by platform authentication in Azure.
    handler: async (request, context) => {
        const requestId = crypto.randomUUID();
        const msRequestId = request.headers.get('x-ms-client-request-id');
        const headerCorrelationId = request.headers.get('x-ms-correlation-id') || null;
        const log = new RequestLog(context, requestId, msRequestId, headerCorrelationId);
        let correlationId = headerCorrelationId || requestId;
        let evaluation = false;
        let httpStatus = 500;
        const respond = (status, jsonBody) => {
            httpStatus = status;
            log.responsePrepared(status, Object.hasOwn(jsonBody, 'nonce'), Object.hasOwn(jsonBody, 'correlationId'));
            return { status, jsonBody };
        };

        try {
            log.service('request_received');
            const config = readConfig();
            const clientRequestId = msRequestId || requestId;

            let payload;
            try {
                payload = JSON.parse(await request.text());
            } catch {
                log.failure('request_validation', 'invalid JSON body', 400);
                return respond(400, { error: 'bad_request', reason: 'invalid JSON body', requestId });
            }

            const parsed = parseEnvelope(payload);
            if (parsed.error) {
                log.failure('request_validation', parsed.error, 400);
                return respond(400, { error: 'bad_request', reason: parsed.error, requestId });
            }
            const envelope = parsed.envelope;
            correlationId = envelope.correlationId || headerCorrelationId || requestId;
            evaluation = envelope.mode === MODE.EVALUATION;
            log.envelopeValidated(envelope, envelope.correlationId || headerCorrelationId,
                envelope.correlationId ? 'envelope' : 'header');

            let delivery;
            let header;
            try {
                ({ context: delivery, header } = await decryptDeliveryContext(
                    envelope.encryptedDeliveryContext, config));
            } catch {
                log.failure('decryption', 'decryption_failed', 400);
                return respond(400, { error: 'decryption_failed', correlationId, requestId });
            }
            log.service('delivery_context_decrypted');

            // Key ID is advisory after authenticated decryption.
            if (config.expectedKeyId && config.expectedKeyId !== header.kid) {
                log.keyIdMismatch();
            }

            if (!delivery?.isComplete) {
                log.failure('delivery_context_validation', 'incomplete delivery context', 400);
                return respond(400, { error: 'bad_request', reason: 'incomplete delivery context', correlationId, requestId });
            }

            // Evaluation proves decryption without resolving a provider or requiring provider config.
            if (!evaluation) {
                const dispatch = contextToDispatch(delivery, envelope, clientRequestId);
                dispatch.correlationId = correlationId;
                const result = await dispatchOtp(dispatch, { requestId, config, log }).catch(() => {
                    if (!log.data.failureStage) log.failure('provider_dispatch', 'unexpected_error', 500);
                    return { httpStatus: 500 };
                });
                if (result.httpStatus !== 200) {
                    return respond(result.httpStatus, { error: 'provider_delivery_failed', correlationId, requestId });
                }
            } else {
                log.service('evaluation_completed');
            }

            // Only a successful delivery (or validated evaluation) may echo the nonce and accepted.
            return respond(200, { nonce: delivery.nonce, correlationId, providerStatus: 'accepted' });
        } catch {
            if (!log.data.failureStage) log.failure('handler', 'unexpected_error', 500);
            return respond(500, { error: 'delivery_failed', correlationId, requestId });
        } finally {
            log.complete(httpStatus);
        }
    },
});
